### ISSUE TITLE
How to refresh asynchronous cache?

### ISSUE THREAD

**Issue by AleksanderBrzozowski** [ben-manes/caffeine#1291]
Hi, I have a question related to cache refresh. Let's say that we have an instance of `AsyncLoadingCache`. In order to refresh a key in this cache, we can use such function:

```java
cache.synchronous().refresh("key").join();
```

But the thing is that sometimes cache gets refreshed, and sometimes not. Below there is a code that shows the problem, we expect that after refresh, if we get the actual version of `key` from cache, we will get a new value instead of the old one. It seems to be nondeterministic - sometimes it returns old value, and sometimes new, refreshed value (see code below).

```java
public class Main {

    public static void main(String[] args) {
        var singleThreadExecutor = Executors.newSingleThreadExecutor();
        IntStream.rangeClosed(0, 100).forEach(value -> {
            var map = new HashMap<String, String>();
            var cache = Caffeine.newBuilder()
                    .executor(singleThreadExecutor)
                    .<String, String>buildAsync((key, executor) -> CompletableFuture.supplyAsync(() -> map.get(key)));
            cache.synchronous().put("b", "b1");
            map.put("b", "b2");
            cache.synchronous().refresh("b").join();
            var cachedValue = cache.get("b").join();
            if (!cachedValue.equals("b2")) {
                System.out.println("Err, iteration: " + value); // here we log if the value is equal to old value
            }
        });
        singleThreadExecutor.shutdown();
    }
}
```

**Comment by ben-manes** [comment]
You have discovered the quirkiness of `CompletableFuture` and the answers won't be very satisfying.

Caffeine attaches a `whenComplete` synchronous dependent action to the future that you supply above in order to update the mapping. The source future is returned by the `refresh` method so when the join completes, the dependent action may not have completed yet due to the non-deterministic nature of thread scheduling.

You might ask then for the dependent action to be returned instead, as you want both the load and the cache update to be completed. The problem is that methods like `cancel` are independent on each future, so you would impact only the cache's dependent action and not the source future or other dependents you placed on it. The cancel would halt our dependent action from running, which could lead to the cache being left in an inconsistent state as its callback was never invoked. A mitigation to safeguard the cache update would be to use further `copy()`, so that canceling has no upstream impact. That still means that the source is not canceled or any of its dependents, nor are the other lifecycle methods usable, so it reduces the implied functionality.

Caffeine tries to return the source future whenever possible as those have reference identity. This way if you have a secondary mapping or use the `asMap()` view's atomic compute/remove/replace methods, any conditional lookups will match on that identity comparison. This way the cache mirrors your expectation for a `Map<K, CompletableFuture<V>>` in all semantics. If you attach lifecycles, like `orTimeout` or complete it out-of-band, it behaves as an unbounded map would. That makes it seem like an extension to the Collection Framework with fewer surprises and just additional features like bounding policies on top of a `ConcurrentHashMap`.

We run into similar challenges when users have dependent actions on the future returned by `AsyncCache.get(key)`. They might expect their action to be able to look back into the cache to see a side effect, like if it failed that the future was removed automatically. Since dependent actions are run in LIFO order, their additional actions run before the cache's. Similarly if we returned a different future then lifecycle and any reliance on identity equality would fail. 

In summary, we are stuck with bad outcomes and the current behavior is the less worse. If `CompletableFuture` added a new join type, "join and wait for dependencies to complete", then you would get the desired behavior. That feature request would have to go to the [Java core libs team](https://mail.openjdk.org/mailman/listinfo/core-libs-dev).

**Comment by AleksanderBrzozowski** [comment]
Thanks for detailed response, @ben-manes 

Actually, this behavior doesn't seem to be a big deal for us, and in our case, the eventual consistency behavior of the refresh method is ok.

I've just wanted to make sure that this is expected behavior, because we encountered it in our integration tests. We have a work-around in tests where we use [awaitility](https://github.com/awaitility/awaitility) library to await certain condition 🙂

**Comment by ben-manes** [comment]
Wonderful, thanks for reporting in case it had been a bug.
